drivers/display/gd7965.c

```diff
@@ -0,0 +1,464 @@
+		return -EINVAL;
+	}
+
+	/* Setup Partial Window and enable Partial Mode */
+	sys_put_be16(x, &ptl[GD7965_PTL_HRST_IDX]);
+	sys_put_be16(x_end_idx, &ptl[GD7965_PTL_HRED_IDX]);
+	sys_put_be16(y, &ptl[GD7965_PTL_VRST_IDX]);
+	sys_put_be16(y_end_idx, &ptl[GD7965_PTL_VRED_IDX]);
+	ptl[sizeof(ptl) - 1] = GD7965_PTL_PT_SCAN;
+	LOG_HEXDUMP_DBG(ptl, sizeof(ptl), "ptl");
+
+	gd7965_busy_wait(driver);
+	if (gd7965_write_cmd(driver, GD7965_CMD_PTIN, NULL, 0)) {
+		return -EIO;
+	}
+
+	if (gd7965_write_cmd(driver, GD7965_CMD_PTL, ptl, sizeof(ptl))) {
+		return -EIO;
+	}
+
+	/* Disable boarder output */
+	bdd_polarity |= GD7965_CDI_BDZ;
+	if (gd7965_write_cmd(driver, GD7965_CMD_CDI,
+			     &bdd_polarity, sizeof(bdd_polarity))) {
+		return -EIO;
+	}
+
+	if (gd7965_write_cmd(driver, GD7965_CMD_DTM2, (u8_t *)buf, buf_len)) {
+		return -EIO;
+	}
+
+	/* Update partial window and disable Partial Mode */
+	if (blanking_on == false) {
+		if (gd7965_update_display(dev)) {
+			return -EIO;
+		}
+	}
+
+	/* Enable boarder output */
+	bdd_polarity &= ~GD7965_CDI_BDZ;
+	if (gd7965_write_cmd(driver, GD7965_CMD_CDI,
+			     &bdd_polarity, sizeof(bdd_polarity))) {
+		return -EIO;
+	}
+
+	if (gd7965_write_cmd(driver, GD7965_CMD_PTOUT, NULL, 0)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int gd7965_read(const struct device *dev, const u16_t x, const u16_t y,
+		       const struct display_buffer_descriptor *desc, void *buf)
+{
+	LOG_ERR("not supported");
+	return -ENOTSUP;
+}
+
+static void *gd7965_get_framebuffer(const struct device *dev)
+{
+	LOG_ERR("not supported");
+	return NULL;
+}
+
+static int gd7965_set_brightness(const struct device *dev,
+				 const u8_t brightness)
+{
+	LOG_WRN("not supported");
+	return -ENOTSUP;
+}
+
+static int gd7965_set_contrast(const struct device *dev, u8_t contrast)
+{
+	LOG_WRN("not supported");
+	return -ENOTSUP;
+}
+
+static void gd7965_get_capabilities(const struct device *dev,
+				    struct display_capabilities *caps)
+{
+	memset(caps, 0, sizeof(struct display_capabilities));
+	caps->x_resolution = EPD_PANEL_WIDTH;
+	caps->y_resolution = EPD_PANEL_HEIGHT;
+	caps->supported_pixel_formats = PIXEL_FORMAT_MONO10;
+	caps->current_pixel_format = PIXEL_FORMAT_MONO10;
+	caps->screen_info = SCREEN_INFO_MONO_MSB_FIRST | SCREEN_INFO_EPD;
+}
+
+static int gd7965_set_orientation(const struct device *dev,
+				  const enum display_orientation
+				  orientation)
+{
+	LOG_ERR("Unsupported");
+	return -ENOTSUP;
+}
+
+static int gd7965_set_pixel_format(const struct device *dev,
+				   const enum display_pixel_format pf)
+{
+	if (pf == PIXEL_FORMAT_MONO10) {
+		return 0;
+	}
+
+	LOG_ERR("not supported");
+	return -ENOTSUP;
+}
+
+static int gd7965_clear_and_write_buffer(struct device *dev,
+					 u8_t pattern, bool update)
+{
+	struct display_buffer_descriptor desc = {
+		.buf_size = GD7965_NUMOF_PAGES,
+		.width = EPD_PANEL_WIDTH,
+		.height = 1,
+		.pitch = EPD_PANEL_WIDTH,
+	};
+	u8_t *line;
+
+	line = k_malloc(GD7965_NUMOF_PAGES);
```

Comment:
`k_malloc` is used but dependency is missing on kernel heap (`HEAP_MEM_POOL_SIZE` is by default 0)

Reply:
We can not use CFB nor LVGL without HEAP, where is the problem?

Reply:
What if the user does not want to use either CFB or LVGL?

Further LVGL can perfectly work without `k_malloc` and use libc heap via `malloc` or a dedicated (user space) memory pool.

I don't see any issue to specify this in the Kconfig to make clear to the users what the actual dependencies are for this driver, or did I miss something?

Reply:
> What if the user does not want to use either CFB or LVGL?

Then initialization would fail, what is okay. I could change the error handling of gd7965_clear_and_write_buffer() and do not treat ENOMEM as an error but as warning.
`depends on HEAP_MEM_POOL_SIZE != 0` is an option, I can add it if you insist :smile:

Reply:
The `ENOMEM` is fine for me.

I thought it would just be nice to add `depends on HEAP_MEM_POOL_SIZE != 0` so that it is clear that there is a dependency.

Reply:
fixed